src/oni2-keyboard-layout/Oni2_KeyboardLayout.re

```diff
@@ -0,0 +1,45 @@
+type callback = (~language: string, ~layout: string) => unit;
+
+type callbackPriv = (string, string) => unit;
+
+let wrapCallback = (callback, language, layout) =>
+  callback(~language, ~layout);
+
+let callbackListRef: ref(list(callbackPriv)) = ref([]);
+
+let subscribe = (callback: callback) =>
+  callbackListRef := [wrapCallback(callback), ...callbackListRef^];
+
+external init: unit => unit = "oni2_KeyboardLayoutInit";
+
+external getCurrentLayout: unit => string =
+  "oni2_KeyboardLayoutGetCurrentLayout";
+external getCurrentLanguage: unit => string =
+  "oni2_KeyboardLayoutGetCurrentLanguage";
+
+Callback.register("oni2_KeyboardLayoutCallbackListRef", callbackListRef);
+
+module Keymap = {
+  type entry = {
+    unmodified: string,
+    withShift: string,
+    withAltGraph: string,
+    withAltGraphShift: string,
+  };
+
+  type t = Hashtbl.t(string, entry);
+
+  external populateCurrentKeymap: (t, (t, string, entry) => unit) => unit =
+    "oni2_KeyboardLayoutPopulateCurrentKeymap";
```

Comment:
It seems like this is more general than `Hashtbl.t` - it could really work for any data structure + a function to 'accumulate' values, like:

```
external populateCurrentKeyMap: ('keymap, ('keymap, string, entry) => unit) => unit
```

Where the `'keymap` type is polymorphic. Just means it is a bit more loosely coupled. The `getCurrent` below could remain unchanged, though.

Reply:
That's a great idea! Added in 216398e!